src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java

```diff
@@ -218,21 +216,27 @@
             MH_BASE_ADDRESS = lookup.findVirtual(MemorySegment.class, "address",
                     methodType(MemoryAddress.class));
             MH_COPY_BUFFER = lookup.findStatic(Binding.class, "copyBuffer",
-                    methodType(MemorySegment.class, MemorySegment.class, long.class, long.class, NativeScope.class));
-            MH_ALLOCATE_BUFFER = lookup.findStatic(MemorySegment.class, "allocateNative",
-                    methodType(MemorySegment.class, long.class, long.class));
+                    methodType(MemorySegment.class, MemorySegment.class, long.class, long.class, SharedUtils.Allocator.class));
+            MH_ALLOCATE_BUFFER = lookup.findStatic(Binding.class, "allocateBuffer",
+                    methodType(MemorySegment.class, long.class, long.class, SharedUtils.Allocator.class));
+            MH_TO_SEGMENT = lookup.findStatic(Binding.class, "toSegment",
+                    methodType(MemorySegment.class, MemoryAddress.class, long.class));
         } catch (ReflectiveOperationException e) {
             throw new RuntimeException(e);
         }
     }
 
     enum Tag {
-        MOVE,
-        DEREFERENCE,
+        VM_STORE,
```

Comment:
Should these be called REG_STORE and REG_LOAD? Ah, I guess we also have stack, right? Ok, I agree with the name - these are "moves" that only the VM can perform, makes sense.

Reply:
REG_LOAD and REG_STORE also works for me. The name is derived from VMReg in the HotSpot code -> VMStorage in our code -> VM_LOAD/VM_STORE. (So, REG works equally well).